http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/sam/SamAuthProvider.scala

```diff
@@ -102,30 +89,40 @@ class SamAuthProvider(val config: Config, serviceAccountProvider: ServiceAccount
       // tokenExpiresIn should not taken into account when comparing cache keys
-      notebookAuthCache.get(NotebookAuthCacheKey(internalId, userInfo.copy(tokenExpiresIn = 0), action, googleProject, clusterName, executionContext))
+      Effect[F].delay(notebookAuthCache.get(NotebookAuthCacheKey(internalId, authorization, action, googleProject, clusterName)).booleanValue())
     } else {
-      hasNotebookClusterPermissionInternal(internalId, userInfo, action, googleProject, clusterName)
+      hasNotebookClusterPermissionInternal(internalId, authorization, action, googleProject, clusterName)
     }
   }
 
-  private def hasNotebookClusterPermissionInternal(internalId: ClusterInternalId, userInfo: UserInfo, action: NotebookClusterActions.NotebookClusterAction, googleProject: GoogleProject, clusterName: ClusterName)(implicit executionContext: ExecutionContext): Future[Boolean] = {
-    // if action is connect, check only cluster resource. If action is anything else, either cluster or project must be true
-    retryUntilSuccessOrTimeout(shouldInvalidateSamCacheAndRetry, s"SamAuthProvider.hasNotebookClusterPermissionInternal call failed for ${googleProject.value}/${clusterName.value}")(samRetryInterval, samRetryTimeout) { () =>
-      Future {
-        val hasNotebookAction = blocking(samClient.hasActionOnNotebookClusterResource(internalId, userInfo, getNotebookActionString(action)))
-        if (action == ConnectToCluster) {
-          hasNotebookAction
-        } else {
-          hasNotebookAction || blocking(samClient.hasActionOnBillingProjectResource(userInfo, googleProject, getProjectActionString(action)))
-        }
+  private def hasNotebookClusterPermissionInternal(internalId: ClusterInternalId, authorization: Authorization, action: NotebookClusterActions.NotebookClusterAction, googleProject: GoogleProject, clusterName: ClusterName)(implicit ev: ApplicativeAsk[F, TraceId]): F[Boolean] = {
+    for {
+      traceId <- ev.ask
+      hasNotebookAction <- hasNotebookClusterPermissionInternal(internalId, action, authorization)
+      res <- if(action == ConnectToCluster) Sync[F].pure(hasNotebookAction) else {
+        if(hasNotebookAction)
+          Sync[F].pure(true)
+        else
+          hasProjectPermissionInternal(googleProject, action, authorization).recover { case _ => false}
       }
-    }
+    } yield res
   }
 
+  private def hasProjectPermissionInternal(googleProject: GoogleProject, action: LeoAuthAction, authHeader: Authorization)(implicit ev: ApplicativeAsk[F, TraceId]): F[Boolean] = for {
+    actionString <- Effect[F].fromEither(getProjectActionString(action))
+    res <- samDao.hasResourcePermission(googleProject, actionString, ResourceTypeName.BillingProject, authHeader)
+  } yield res
+
+  private def hasNotebookClusterPermissionInternal(clusterInternalId: ClusterInternalId, action: LeoAuthAction, authHeader: Authorization)(implicit ev: ApplicativeAsk[F, TraceId]): F[Boolean] = for {
```

Comment:
Do we need this overloaded method or can it just be folded into the other `hasNotebookClusterPermissionInternal`?

Reply:
nice catch! I definitely didn't mean to overload the function. They can't really be folded into one another because they're reused in different places, so I just renamed one of them